tools/base/libboard/board_rc/board_rc.c

```diff
@@ -0,0 +1,291 @@
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <glob.h>
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+#include <opae/properties.h>
+#include <opae/utils.h>
+#include <opae/fpga.h>
+
+#include "safe_string/safe_string.h"
+#include "board_rc.h"
+
+// BMC sysfs path
+#define SYSFS_DEVID_FILE "avmmi-bmc.*.auto/bmc_info/device_id"
+#define SYSFS_RESET_FILE "avmmi-bmc.*.auto/bmc_info/reset_cause"
+#define SYSFS_PWRDN_FILE "avmmi-bmc.*.auto/bmc_info/power_down_cause"
+
+#define FPGA_STR_SIZE   256
+
+// read bmc version
+fpga_result read_bmc_version(fpga_token token, int *version)
+{
+	fpga_result res               = FPGA_OK;
+	fpga_result resval            = FPGA_OK;
+	device_id dev;
+	fpga_object bmc_object;
+	fpga_handle handle;
+
+	if (version == NULL) {
+		OPAE_ERR("Invalid input parameter");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaOpen(token, &handle, FPGA_OPEN_SHARED);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to open fpga device");
+		resval = res;
+		goto out;
+	}
+
+	res = fpgaHandleGetObject(handle, SYSFS_DEVID_FILE, &bmc_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get handle Object");
+		resval = res;
+		goto out_close;
+	}
+
+	res = fpgaObjectRead(bmc_object, (uint8_t*)(&dev), 0, sizeof(dev), 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to Read object ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	*version = dev.aux_fw_rev_0_7
+		| (dev.aux_fw_rev_8_15 << 8)
+		| (dev.aux_fw_rev_16_23 << 16)
+		| (dev.aux_fw_rev_24_31 << 24);
+
+
+out_destroy:
+	res = fpgaDestroyObject(&bmc_object);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to Destroy Object");
+		resval = res;
+	}
+
+out_close:
+	res = fpgaClose(handle);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to close handle");
+		resval = res;
```

Comment:
resval is overwritten, losing the previous error if any.

Reply:
Same comment applies to the goto logic in the remaining fns.

Reply:
fixed